### ISSUE TITLE
Running dotnet tests in parellel throws error - it is being used by another process. 

### ISSUE THREAD

**Issue by vijayrkn** [dotnet/cli#2616]
## Steps to reproduce

discovering multiple tests in parallel causes this error
C:\Program Files\dotnet\dotnet.exe compile-csc @E:\git\mvc\src\Microsoft.AspNetCore.Mvc.Abstractions\obj\Debug\netstandard1.3\dotnet-compile.rsp returned Exit Code 1
## Expected  behavior

Building in parallel works.
## Actual behavior

The process cannot access the file 'E:\git\mvc\src\Microsoft.AspNetCore.Mvc.Abstractions\obj\Debug\netstandard1.3\dotnet-resgen-resx.rsp' because it is being used by another process.
## Environment data

`dotnet --info` output:

Product Information:
 Version:     1.0.0-rc2-002439
 Commit Sha:  ef0c3b2cee

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64

**Comment by piotrpMSFT** [comment]
@cdmihai it looks like VS is trying to do multiple builds concurrently and the builds are conflicting with one another. Forgetting `dotnet test`, this means that running multiple builds in parallel is not handled gracefully.

**Comment by cdmihai** [comment]
OK. We can use OS level locks to coordinate between processes. Here's one solution:

- locking mechanism: whatever each OS provides. Hopefully file handle's FileShare.None works as a mutex crossplat. If not, we'll have OS specific locking code
- locking granularity: per config / tfm combination (each output path gets a lock)
- locking rule: 
1) build wants to build a project; 
2) build waits to acquire the OS lock for the output path
3) build either had incremental hit or calls csc
4) build releases OS lock for project

Regarding deadlocks, there shouldn't be any. There's no reentrant lock acquisition, each process takes only one lock at a time and releases it.

**Comment by cdmihai** [comment]
@vijayrkn I'd like to have this for RTM, not RC2. Is that alright?

**Comment by cdmihai** [comment]
Beyond test and build, should other commands behave "gracefully" when called in parallel from other processes? Publish, pack ..?

**Comment by vijayrkn** [comment]
yes RTM should be fine because we have decided to call dotnet test with --no-build flag (from VS) for RC2. We need to fix this for dotnet publish too.

**Comment by piotrpMSFT** [comment]
bumping to RTM per above.

**Comment by sayedihashimi** [comment]
It seems like a really bad user experience for the test cases to be running old code. I would think that this would be super frustrating for users. They are editing test code to get the test to pass, but the test explorer is using old code so the test never passes until an explicit build.

I discussed with @barrytang @mlorbetske and we are hoping that we can fix this somehow for RC2, or in a microupdate somehow.

**Comment by barrytang** [comment]
@piotrpMSFT @Petermarcu, this is the issue I mentioned in Core War today. Now you understand the tooling impact, are you guys still thinking this is out of scope for RC2?

**Comment by cdmihai** [comment]
There's also the question whether this (graceful handling when called in parallel from other processes) is CLI's responsibility. For example, MSBuild crashes similarly when multiple invocations race on the same files.

In CLI's case, the processes that invoke dotnet build can ensure they send unique build base paths per invocation, thus eliminating the race (it also eliminates sharing build artifacts unfortunately).

**Comment by vijayrkn** [comment]
VS currently passing the same build-base-path for all projects while invoking dotnet test and dotnet test constructs the full output path based on configuration, framework and rid. Are you suggesting that dotnet test (or any other tool that calls build) always constructs unique path for the build outputs while invoking build? Wouldn't we loose the benefit of incremental build in this case?

**Comment by cdmihai** [comment]
Since VS is calling dotnet test in parallel, it should ensure that each dotnet test invocation has a unique build base path. You loose sharing of build artefacts between invocations. Within invocations though, the artefacts are still being reused.

BTW, are you using a --no-dependency argument for dotnet test? If not, dotnet build rebuilds the entire project graph (reachable from the test project). It is my understanding that VS always uses the --no-dependency flag because msbuild is the dependency orchestrator. Is that different for test invocations?

**Comment by vijayrkn** [comment]
The current behavior is passing the --no-build and passing the output path (containing the dlls) to dotnet test and dotnet test discovers/runs the test from this provided output path. 

But we have a bug where - if the user changes anything in the test and triggers a test execution, project build/msbuild is not triggered at this point. So, the user could get wrong test results.

That is the reason we were exploring the option not to pass --no-build to dotnet test. This way dotnet test would always build the latest before executing the test. 

The other option is for VS to always trigger a project build before triggering dotnet test (we are testing this option too. If this works, we wont need to pass the build option to dotnet test). In this case, we wont need this bug fixed.

To answer the --no-dependencies  question, we do not pass the --no-dependencies  option to dotnet test currently (I didn't see this option in dotnet test --help).

**Comment by cdmihai** [comment]
@vijayrkn You are right, I also checked and dotnet test does not have a --no-dependencies flag to pass thru to build.

Triggering a build of the test project when the test project changed, from within VS, seems to me like the good approach here, since VS is the coordinator.

**Comment by vijayrkn** [comment]
We verified the feasibility of solution (1) and we should be able to trigger a VS build (when ever there is a change in the project) before invoking dotnet test. We can go with this solution for rc2.

I think we should still fix (2), adding an option in dotnet test to pass --no-dependencies.

Regarding (3), I am not sure if we can enforce all callers to make sure they pass unique paths for base path.

Shouldn't we still fix the root cause of this issue for RTW  - which is enabling parallel build to same the base path?

**Comment by piotrpMSFT** [comment]
This is a project.json-specific bug. We don't have plans to ship additional project.json updates now that we've transitioned to msbuild. Closing the issue, therefore.
